### ISSUE TITLE
`@typeName(T)` should be equal to itself

### ISSUE THREAD

**Issue by leecannon** [ziglang/zig#13961]
### Zig Version

0.11.0-dev.811+8ff9284c4

### Steps to Reproduce and Observed Behavior

```zig
pub fn main() void {
    @compileLog(@typeName(u8) == @typeName(u8));
}
```
```
$ zig build run
main.zig:2:5: error: found compile log statement
    @compileLog(@typeName(u8) == @typeName(u8));
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Compile Log Output:
@as(bool, false)
error: fff...
error: The following command exited with error code 1:
/home/lee/zig/0.11.0-dev.811+8ff9284c4/files/zig build-exe /home/lee/src/fff/main.zig --cache-dir /home/lee/src/fff/zig-cache --global-cache-dir /home/lee/.cache/zig --name fff --enable-cache
error: the following build command failed with exit code 1:
/home/lee/src/fff/zig-cache/o/f4c4dd9752aa2a2914a9207a728e83d8/build /home/lee/zig/0.11.0-dev.811+8ff9284c4/files/zig /home/lee/src/fff /home/lee/src/fff/zig-cache /home/lee/.cache/zig run
```

### Expected Behavior

`@typeName(u8) == @typeName(u8)` to be true.

**Comment by r00ster91** [comment]
Does this make sense? `@typeName` returns a pointer. AFAIK pointers don't really exist at comptime right now and there are things like #11734. I wonder if this comparison should be allowed at all?
At runtime, in debug mode, the addresses seem to be different:
```zig
const std = @import("std");

pub fn main() void {
    std.debug.print("{*}\n{*}\n", .{@typeName(u8), @typeName(u8)});
}
```
```
$ zig run x.zig # debug mode
[2:0]u8@208c84
[2:0]u8@208c90
```
So maybe it should be false. However, the pointers can also point to the same string with optimizations on:
```
$ zig run x.zig -O ReleaseSmall
[2:0]u8@202c45
[2:0]u8@202c45
```
So, should the result of this depend on the build mode? Well I think perhaps the comparison should be disallowed because pointers AFAIK don't really exist right now at comptime. You should use `std.mem.eql` if you mean to compare the characters.

**Comment by nektro** [comment]
so far one of the rules of comptime is that comptime calls are memoized (eg calls made with the same value arguments, return the same result as a previous call), so this is a fairly reasonable expectation imo

the note about `ReleaseSmall` is likely llvm performing string interning of constants

**Comment by leecannon** [comment]
The unexpected nature of this highlighed below:
```zig
pub fn main() void {
    @compileLog(@typeName(u8) == @typeName(u8));
    @compileLog(comptime myTypeName(u8) == myTypeName(u8));
}

fn myTypeName(comptime T: type) *const [lengthOfTypeName(T)]u8 {
    return @typeName(T);
}

fn lengthOfTypeName(comptime T: type) comptime_int {
    return @typeName(T).len;
}
```
It is weird that a different result is reached when using a trivial wrapper around the builtin.

The fact that `myTypeName` is memoized so multiple calls with the same arguments are equal but the same is not true of `@typeName` is not something that would be expected.

**Comment by nektro** [comment]
`@typeName` is a builtin so its handled very differently from normal functions in the compiler

**Comment by zzyxyzz** [comment]
> `@typeName` is a builtin so its handled very differently from normal functions in the compiler

Yes, but this flexibility should not be used gratuitously. Like you said yourself, in this case it is reasonable to expect that `@typeName` is memoized at comptime like every other function. 

Comptime semantics is general is quite a big enough mess already. A bit of consistency here and there could do it some good.

**Comment by tecanec** [comment]
That `@typeName(T) != @typeName(T)` when the strings are stored seperately doesn't surprice me. What does surprice me is that there's no deduplication going on.

Does string deduplication only happen before comptime?

**Comment by SuperAuguste** [comment]
This is now resolved (tested on `0.12.0-dev.3635+786876c05`).
